src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -479,17 +522,47 @@ BookStep<TIn, TOut, TDerived>::qualityUpperBound(
 {
     auto const dir = this->debtDirection(v, StrandDirection::forward);
 
-    // This can be simplified (and sped up) if directories are never empty.
-    Sandbox sb(&v, tapNONE);
-    BookTip bt(sb, book_);
-    if (!bt.step(j_))
+    std::optional<std::pair<Quality, bool>> const res = tipOfferQuality(v);
```

Comment:
If we just look at thiks code locally, what's the meaning of the second value of the pair? The only clues we get are: it's a bool, it's returned from `tipOfferQuality` and it's the last parameter to `adjustQualityWithFees`'. It's almost impossible to figure out what that is from local information. Consider replacing the `bool` with an `enum`.